sdk/go/internal/engineconn/session.go

```diff
@@ -18,15 +19,19 @@ import (
 type cliSessionConn struct {
 	*http.Client
 	childStdin io.Closer
+	childProc  *exec.Cmd
```

Comment:
If we go with this solution we should ensure that the weird problem from before where GC was closing the stdin pipe is gone, had a way of reproing it here: https://github.com/dagger/dagger/pull/4551#issuecomment-1421939344. Even though we now include childProc in this struct, before we were including childStdin in this struct but the pipe it references still seemed to be gc'd. It didn't really make total sense at the time so it's worth a double check.
* While I agree that users should be calling `Close`, the problem is that if they forgot to then the GC weirdness resulted in a totally incomprehensible error rather than just a simple resource leak.

However, if the problem is around Pdeathsig, then it may be even simpler to just remove that and leave everything else as is. It doesn't work on non-Linux platforms and was mostly left in because this code was originally based on some other code that included Pdeathsig. The approach of having the child process exit when its stdin is closed works on at least Linux+MacOS and should be sufficient I think. I'm not even sure if the other SDKs set Pdeathsig.

Reply:
Actually I just saw the CI for this commit and it does in fact contain the EOF error that the `go proc.Wait()` fixed previously: https://github.com/dagger/dagger/actions/runs/4257804458/jobs/7408340327#step:6:1500

So yeah if getting rid of Pdeathsig and leaving the rest solves the log flush problem then I think it's gonna be the simplest one.

Reply:
The plot thickens! Thanks for the pointers, that was definitely a curious failure.

I'll try getting rid of the Pdeathsig.

Reply:
Just to clarify what I meant before, you'll need to get rid of Pdeathsig and also go back to the go proc.Wait() from before in order to fix the EOF failure consistently I think.

Reply:
I don't think we can just remove the `Pdeathsig` and have the single `go proc.Wait()` because that would leave a gap where the calling program exits but the background session process is still writing logs. It's better than killing the session process, but still a race since whatever ran the calling program might immediately start processing the logs.

I'll run your repro command and see if I happened to fix the EOF somehow with the rest of these changes, since I haven't seen it in the past 5 actions runs or so. If it's still happening I'll switch to an `errgroup.Group` so we always `Wait` and then change `Close` to just wait on the errgroup, since we can't call `(*exec.Cmd).Wait` twice.

Reply:
Ran the repro for 45 minutes until one of the runs somehow consumed all memory on my machine. I'm pretty sure that was unrelated, I've seen it happen in other situations but I'm not close to understanding what triggers it. :thinking:

Reply:
> I don't think we can just remove the Pdeathsig and have the single go proc.Wait() because that would leave a gap where the calling program exits but the background session process is still writing logs. 

Ah right I forgot that we have an intermediary pipe setup by the SDK process.

> If it's still happening I'll switch to an errgroup.Group so we always Wait and then change Close to just wait on the errgroup, since we can't call (*exec.Cmd).Wait twice.

SGTM

> Ran the repro for 45 minutes until one of the runs somehow consumed all memory on my machine. I'm pretty sure that was unrelated, I've seen it happen in other situations but I'm not close to understanding what triggers it. 🤔

Yeah that's not ideal, if you can repro again it would be interesting to see if it's the sdk process, the cli subprocess or the engine running in docker.

Reply:
Moving this into #4656